pydantic/main.py

```diff
@@ -1094,18 +1095,33 @@ def _calculate_keys(self, *args: Any, **kwargs: Any) -> Any:
 RootModelRootType = typing.TypeVar('RootModelRootType')
 
 
+class _RootModelSpecialValues(enum.Enum):
+    no_value = enum.auto()
+
+
+_RootModelNoValue = _RootModelSpecialValues.no_value
+
+
 class RootModel(BaseModel, typing.Generic[RootModelRootType]):
     __pydantic_root_model__ = True
-    # TODO: Make `__pydantic_fields_set__` logic consistent with `BaseModel`, i.e. it should be `set()` if default value
-    # was used
-    __pydantic_fields_set__ = {'root'}  # It's fine having a set here as it will never change
     __pydantic_private__ = None
     __pydantic_extra__ = None
 
     root: RootModelRootType
 
-    def __init__(__pydantic_self__, root: RootModelRootType) -> None:  # type: ignore
+    def __init__(
+        __pydantic_self__, root: RootModelRootType | Literal[_RootModelNoValue] = _RootModelNoValue
```

Comment:
I don't love that this exposes a private type in the signature, it's annoying if you want to subclass and override the init and keep the signature the same. But maybe this is the best alternative. @samuelcolvin I do think it might be nice to make `pydantic._internal._fields.Undefined` public for this and other use cases (such as introducing a "blessed" way to cause fields to not get serialized). I guess we can do that later though.